off_chain_data/application-go/parser/block.go

```diff
@@ -0,0 +1,101 @@
+package parser
+
+import (
+	"fmt"
+
+	"github.com/hyperledger/fabric-protos-go-apiv2/common"
+	"google.golang.org/protobuf/proto"
+)
+
+type Block struct {
+	block              *common.Block
+	cachedTransactions []*Transaction
+}
+
+func ParseBlock(block *common.Block) *Block {
+	return &Block{block, nil}
+}
+
+func (b *Block) Number() uint64 {
+	return b.block.GetHeader().GetNumber()
+}
+
+func (b *Block) Transactions() ([]*Transaction, error) {
+	if b.cachedTransactions != nil {
+		return b.cachedTransactions, nil
+	}
+
+	funcName := "Transactions"
+	envelopes, err := b.unmarshalEnvelopesFromBlockData()
+	if err != nil {
+		return nil, fmt.Errorf("in %s: %w", funcName, err)
+	}
+
+	commonPayloads, err := b.unmarshalPayloadsFrom(envelopes)
+	if err != nil {
+		return nil, fmt.Errorf("in %s: %w", funcName, err)
+	}
+
+	payloads, err := b.parse(commonPayloads)
+	if err != nil {
+		return nil, fmt.Errorf("in %s: %w", funcName, err)
+	}
+
+	b.cachedTransactions = b.createTransactionsFrom(payloads)
+
+	return b.cachedTransactions, nil
+}
```

Comment:
I think there might be a simpler way to do the lazy unmarshal and caching of results by using the [sync](https://pkg.go.dev/sync) package from the standard library. This might be worth trying:

```suggestion
type Block struct {
	block        *common.Block
	transactions func() ([]*Transaction, error)
}

func ParseBlock(block *common.Block) *Block {
	result := &Block{
		block: block,
	}
	result.transactions = sync.OnceValues(result.unmarshalTransactions)

	return result
}

func (b *Block) Number() uint64 {
	return b.block.GetHeader().GetNumber()
}

func (b *Block) Transactions() ([]*Transaction, error) {
	return b.transactions()
}

func (b *Block) unmarshalTransactions() ([]*Transaction, error) {
	envelopes, err := b.unmarshalEnvelopesFromBlockData()
	if err != nil {
		return nil, err
	}

	commonPayloads, err := b.unmarshalPayloadsFrom(envelopes)
	if err != nil {
		return nil, err
	}

	payloads, err := b.parsePayloads(commonPayloads)
	if err != nil {
		return nil, err
	}

	return b.createTransactionsFrom(payloads), nil
}
```

Reply:
Yes, this is great. I added it everywhere.